### ISSUE TITLE
Enable .size.json Sidecar-Free Graph-Budget Batching for AC and MCTS Actor-Critic Training

### ISSUE THREAD

**Issue by 123nol** [iCog-Labs-Dev/Mathematics-ai-atp#39]
Branch: AC-branch && MCTS-AC-branch

## Summary

The supervised actor-critic training pipeline supports graph-budget batching
through:

- `training.max_batch_nodes`
- `training.max_batch_edges`

The current batching path requires every prepared PyG graph to have a matching
`.size.json` sidecar. The existing prepared dataset generated by the
dataset-generation branch contains individual `.pt` graphs and aggregate graph
statistics, but does not contain per-graph sidecars.

As a result, enabling graph budgets fails with:

```text
Prepared graph ... has no size sidecar ... Re-run preparation before enabling graph budgets.
```

The same sidecar-only behavior exists on both `AC-branch` and
`MCTS-AC-branch`.

## Why Graph-Budget Batching Is Required

Fixed graph-count batching does not bound the amount of graph data in a batch.
Two batches with the same number of graphs can contain very different numbers
of nodes and edges.

This caused CUDA out-of-memory failures during actor-critic training. One
failure reported:

```text
GPU capacity: 23.56 GiB
Free memory: 5.87 GiB
PyTorch allocated: 12.56 GiB
PyTorch reserved but unallocated: 4.81 GiB
Allocation requested: 6.01 GiB
```

After reducing the graph-count batch size to 32, training still failed after
several hundred batches:

```text
GPU capacity: 23.56 GiB
Free memory: 4.05 GiB
PyTorch allocated: 14.31 GiB
PyTorch reserved but unallocated: 4.88 GiB
Allocation requested: 8.58 GiB
```

The actor-critic model has several sources of memory use:

- GNN activations scale with total nodes.
- Message passing and GAT attention scale with total edges.
- GATv2 heads and layers increase intermediate tensors.
- Pointer argument selection retains node-level embeddings.
- CUDA allocator fragmentation can increase reserved memory.

Graph-budget batching is therefore needed to bound the total nodes and edges
processed by each PyG batch. `batch_size` remains a maximum graph count, while
the node and edge budgets constrain the total graph contents.

Budgets cannot guarantee that every possible OOM condition disappears, but they
bound the main variable that makes fixed graph-count batches unpredictable.

## Root Cause

`PreparedGraphDataset.graph_sizes()` currently:

1. Iterates over prepared `.pt` paths.
2. Looks for `<graph>.size.json`.
3. Reads the configured `edges_<edge_mode>` field.
4. Raises immediately when the sidecar is missing.

The `.pt` loading logic exists separately in `__getitem__()`, which runs only
after the sampler has already been constructed. Therefore, the current code
never falls back to reading graph sizes from the `.pt` files.

The aggregate `graph_stats` values in split manifests are not sufficient for
batching. They contain split-level means and medians, not the exact size of each
graph.

## Packed Cache Overview

A packed cache is an optional derived cache that groups many individual PyG
graph files into larger chunk files. It is stored beside the prepared dataset:

```text
prepared_root/
  train/pyg/*.pt
  val/pyg/*.pt
  test/pyg/*.pt
  packed/
    bidirectional/
      manifest.json
      train/chunk_000000.pt
      val/chunk_000000.pt
      test/chunk_000000.pt
```

Each chunk is a serialized Python list of PyG `Data` objects. The packed-cache
manifest records the split counts and chunk names. The cache is created from
the existing `.pt` files; it does not regenerate the dataset, rebuild
vocabularies, or change split contents.

The edge mode is part of the cache path because the graph stored in a
`bidirectional` cache has a different `edge_index` from a graph stored in a
`forward` cache.

## How Batching Uses the Packed Cache

When graph budgets are enabled, the loader needs one `(nodes, edges)` pair for
every graph before it constructs the sampler.

The planned resolution order is:

1. If all graph sidecars are valid for the configured edge mode, use their
   values without loading graph objects just to calculate sizes.
2. If sidecars are missing or incomplete and a complete packed cache exists,
   load the cache chunks into the dataset's CPU cache.
3. For each cached graph, calculate:

   ```python
   nodes = int(data.num_nodes)
   edges = int(data.edge_index.size(1))
   ```

4. Pass those graph sizes to `GraphBudgetBatchSampler`.
5. The sampler greedily adds graphs to a batch until adding another graph would
   exceed `batch_size`, `max_batch_nodes`, or `max_batch_edges`.
6. The resulting list of graph indices is passed to PyG `DataLoader` through
   `batch_sampler`.

The size calculation must happen after the configured edge transformation. For
`bidirectional` mode, reverse edges are added before the edge count is used for
budgeting. This ensures that the sampler limits the graph actually processed by
the GNN.

If neither valid sidecars nor a usable packed cache exists, the implementation
can scan the individual `.pt` files once and derive sizes using the same
normalization path as `__getitem__()`. This direct scan is exact but can make
loader construction slow and can cause the graphs to be read again during
training. The packed cache is preferred for repeated runs.

The packed cache is a host-memory and I/O optimization. It is not a model
checkpoint and should normally remain beside the prepared dataset rather than
inside a training run directory.

## Proposed Changes

- Add sidecar-first graph-size resolution.
- Add packed-cache loading and validation.
- Add direct `.pt` size scanning when no packed cache is available.
- Preserve the existing sidecar path for newly prepared datasets.
- Reuse the existing `GraphBudgetBatchSampler`.
- Add explicit oversized-graph policies:
  - `error`
  - `skip`
  - `singleton`
- Add `cache_in_memory` configuration support.
- Pass all batching settings through `ActorCriticConfig` and
  `build_dataloaders()`.
- Call `set_epoch()` on the training sampler before each actor-critic epoch.
- Report budget settings, cache mode, oversized graph counts, and batch counts.

## MCTS Branch Scope

Apply the same shared dataset and batching fix to `MCTS-AC-branch`.

Preserve MCTS-specific behavior, including:

- `selection_policy`
- `num_simulations`
- `sim_batch_size`
- HTPS and replay settings
- Pantograph integration
- MCTS search limits

The RL fields `max_update_nodes` and `max_update_edges` remain separate from
PyG training batch budgets. They limit RL transition updates and must not be
routed through `GraphBudgetBatchSampler`.

## Acceptance Criteria

- Datasets with valid sidecars continue using the sidecar path.
- Datasets without sidecars can train without regenerating preparation
  artifacts.
- Existing packed caches are accepted and validated.
- Individual `.pt` files provide a direct fallback when no packed cache exists.
- `max_batch_nodes` and `max_batch_edges` constrain actual transformed PyG
  batches.
- Forward and bidirectional edge modes produce correct edge counts.
- Oversized graphs are handled according to the configured policy.
- Actor-critic pointer fields remain available.
- MCTS-specific search and replay behavior remains unchanged.
- Fixed-count batching continues to work when both graph budgets are zero.
- Focused batching tests and the relevant full test suite pass.

## Suggested Test Coverage

- Valid sidecars are preferred over graph loading.
- Missing sidecars trigger packed-cache resolution.
- Packed-cache counts and chunk files are validated.
- Direct `.pt` scanning works without sidecars or packed chunks.
- Bidirectional edge counts include transformed reverse edges.
- `error`, `skip`, and `singleton` oversized policies behave as documented.
- Actor-critic batches retain `premise_mask` and `arg_node_indices`.
- MCTS actor-critic loading works with a sidecar-free prepared dataset.
- Zero budgets preserve ordinary fixed-count batching.

**Comment by 123nol** [comment]
this commit to the AC-branch has fixed this issue:
84dfc7e6bec371ccdc45b44d9053097f29c9d405
